### PR TITLE
added from_generator method to IterableDataset class.

### PR DESCRIPTION
**Pull request by hamid-vakilzadeh**
Hello, 

This resolves issues #4988.

I added a method `from_generator` to class `IterableDataset`.

I modified the `read` method of input stream generator to also return Iterable_dataset.

### CODE REVIEW

File: src/datasets/io/generator.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ def __init__(
             **kwargs,
         )
 
-    def read(self):
+    def read(self, streaming: bool = False):
```
Comment (mariosasko) [inline src/datasets/io/generator.py:27]: This is a new param, so let's move to the end.

**Review by mariosasko — COMMENTED** [review]
Great job, thanks! 

Besides my comment below, we also need a test (let me know if you need help with that), and then we can merge.

File: src/datasets/io/generator.py (hunk 1 above)
Comment (hamid-vakilzadeh) [reply src/datasets/io/generator.py:27]: Thank you @mariosasko for your feedback.
I am not sure I understand what I need to do here. Can you please help me on this.
Also a little help on generating the test module would be appreciated.
Thank you!

Comment (mariosasko) [reply src/datasets/io/generator.py]: Oops, I didn't notice that this is the only param in the signature 🙂.

**Comment by mariosasko** [comment]
I added a test and moved the `streaming` param from `read` to `__init_`. Then, I also decided to update the `read` method of the rest of the packaged modules to account for this param. 

@hamid-vakilzadeh Are you OK with these changes?

File: src/datasets/io/abc.py [hunk 2]
```diff
@@ -14,17 +16,19 @@ def __init__(
...
         self.split = split if split or isinstance(path_or_paths, dict) else "train"
         self.features = features
         self.cache_dir = cache_dir
         self.keep_in_memory = keep_in_memory
+        self.streaming = streaming
         self.kwargs = kwargs
 
     @abstractmethod
-    def read(self) -> Union[Dataset, DatasetDict]:
+    def read(self) -> Union[Dataset, DatasetDict, IterableDataset, IterableDataset]:
```
Comment (lhoestq) [inline src/datasets/io/abc.py:31]: IterableDatasetDict maybe ?

**Review by lhoestq — APPROVED** [review]
Love it ! Thanks :)

**Pull request merged**